libtsuba/src/RDGSlice.cpp

```diff
@@ -11,6 +11,142 @@
 #include "tsuba/RDGPrefix.h"
 #include "tsuba/RDGTopology.h"
 
+namespace {
+katana::Result<void>
+demand_load_metadata_array(
+    std::string array_name, std::function<void()> empty, tsuba::RDGCore* core) {
+  auto it = std::find_if(
+      core->part_header().part_prop_info_list().begin(),
+      core->part_header().part_prop_info_list().end(),
+      [&array_name](tsuba::PropStorageInfo& psi) {
+        return psi.name() == array_name;
+      });
+
+  if (it == core->part_header().part_prop_info_list().end()) {
+    // caller's function for emptying the array we care about
+    empty();
+    return katana::ResultSuccess();
+  }
+
+  std::vector<tsuba::PropStorageInfo*> prop_infos{&(*it)};
+  KATANA_CHECKED(AddProperties(
+      core->rdg_dir(), tsuba::NodeEdge::kNeitherNodeNorEdge, nullptr, nullptr,
+      prop_infos, nullptr, [&core](const std::shared_ptr<arrow::Table>& props) {
+        return core->AddPartitionMetadataArray(props);
+      }));
+
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+demand_remove_metadata_array(
+    std::string array_name, std::function<void()> empty, tsuba::RDGCore* core) {
+  auto it = std::find_if(
+      core->part_header().part_prop_info_list().begin(),
+      core->part_header().part_prop_info_list().end(),
+      [&array_name](tsuba::PropStorageInfo& psi) {
+        return psi.name() == array_name;
+      });
+  if (it != core->part_header().part_prop_info_list().end()) {
+    it->WasUnloaded();
+  }
+  empty();
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+demand_load_property(
+    const std::string& name, const tsuba::RDGSlice::SliceArg& slice_arg,
+    tsuba::NodeEdge node_edge, tsuba::RDGCore* core) {
+  if (node_edge == tsuba::NodeEdge::kNeitherNodeNorEdge) {
+    return KATANA_ERROR(
+        tsuba::ErrorCode::InvalidArgument,
+        "cannot load property that is attached to neither nodes nor edges");
+  }
+  auto it = std::find_if(
+      core->part_header().part_prop_info_list().begin(),
+      core->part_header().part_prop_info_list().end(),
+      [&name](tsuba::PropStorageInfo& psi) { return psi.name() == name; });
+
+  if (it == core->part_header().part_prop_info_list().end()) {
+    return tsuba::ErrorCode::PropertyNotFound;
+  }
+
+  std::vector<tsuba::PropStorageInfo*> property{&(*it)};
```

Comment:
Interesing. So you're initializing a new vector with an iterator. We can do that?

Reply:
Why are you creating a vector here though? Is that the requirement for `AddPropertySlice`?

Reply:
Yeah I see that now. At a first glance I thought he was initializing a vector based on an iterator that pointed to the beginning of some container and then it just used that iterator to create the rest of the new vector. I see that it's just a single element now.

Reply:
Nit
While I think this is kind of elegant, it requires a bit more thinking than I'd prefer. 
Maybe to add a bit of clarity split this into:
```
tsuba::PropStorageInfo* prop = &(*it);
std::vector<tsuba::PropStorageInfo*> property{prop};
```
The compiler optimizes it all the same according to godbolt.

Reply:
This has been updated so that there is no longer an iterator at all (searching has been factored out into a `RDGCore` method).